crates/transpiler/src/passes/alap_schedule_analysis.rs

```diff
@@ -0,0 +1,175 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::TranspilerError;
+use hashbrown::HashMap;
+use pyo3::prelude::*;
+use pyo3::types::PyDict;
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType, Wire};
+use qiskit_circuit::operations::{OperationRef, StandardInstruction};
+use qiskit_circuit::{Clbit, Qubit};
+use rustworkx_core::petgraph::prelude::NodeIndex;
+
+#[pyfunction]
+/// Runs the ALAPSchedule analysis pass on dag.
+///
+/// Args:
+///     dag (DAGCircuit): DAG to schedule.
+///     clbit_write_latency (u64): The latency to write classical bits.
+///     node_durations (HashMap<usize, f64>): Mapping from node indices to operation durations.
+///
+/// Returns:
+///     PyDict: A dictionary mapping each DAGOpNode to its scheduled start time.
+///
+#[pyo3(name = "alap_schedule_analysis", signature= (dag, clbit_write_latency, node_durations))]
+pub fn run_alap_schedule_analysis(
+    py: Python,
+    dag: &DAGCircuit,
+    clbit_write_latency: u64,
+    node_durations: HashMap<usize, f64>,
```

Comment:
Hmmm... I'm not quite sure using this type should be the best way of going around this. It should be best to use the defined `NodeIndex` type. Though I get you won't be able to directly extract it from Python, but there should be a way around that too...